File: backend/game/game.py
"""The game itself: owns the board, drives the tick, and is the only thing that
decides what is true. The browser renders whatever this produces.
"""

from __future__ import annotations

import random
from enum import Enum
from typing import Any

from .collision import is_fatal
from .snake import Cell, Direction, Snake

DEFAULT_WIDTH = 24
DEFAULT_HEIGHT = 24

#: Bounds on a board the browser asks for. The lower one keeps `reset()` able to
#: lay a three-cell snake out on the row; the upper one stops a hostile client
#: asking for a board big enough to matter.
MIN_DIMENSION = 8
MAX_DIMENSION = 240
DEFAULT_TICK_SECONDS = 0.12

#: How many background/foreground pairs the browser cycles through. The server
#: sends only an *index*; the hex values live in web/lib/palette.ts, so a colour
#: can be retuned without touching or restarting the backend.
PALETTE_COUNT = 6


class GameStatus(str, Enum):
    READY = "ready"
    RUNNING = "running"
    PAUSED = "paused"
    GAME_OVER = "game_over"


class Game:
    """A single game session. One per connected player.

    `tick()` is the whole game in one call: resolve death, resolve food, move.
    Everything else is setup or serialization.
    """

    def __init__(
        self,
        width: int = DEFAULT_WIDTH,
        height: int = DEFAULT_HEIGHT,
        tick_seconds: float = DEFAULT_TICK_SECONDS,
        seed: int | None = None,
    ) -> None:
        self.width = width
        self.height = height
        self.tick_seconds = tick_seconds
        self._rng = random.Random(seed)
        self.reset()

    # --- setup -----------------------------------------------------------

    def reset(self) -> None:
        self.snake = Snake((self.width // 2, self.height // 2), Direction.RIGHT)
        self.food: Cell | None = None
        self._respawn_food()
        self.score = 0
        self.ticks = 0
        self.palette = 0
        self.status = GameStatus.READY

    def resize(self, width: int, height: int) -> None:
        """Reshape the board to what the browser can show, and start over.

        The board fills the viewport, so its shape is the one piece of game
        state only the browser knows. Resizing restarts the game rather than
        trying to rescue a snake that may now be outside the board - a window
        drag mid-run is rare, and a silently teleported snake is worse than an
        obvious fresh start.
        """
        width = min(max(width, MIN_DIMENSION), MAX_DIMENSION)
        height = min(max(height, MIN_DIMENSION), MAX_DIMENSION)
        if width == self.width and height == self.height:
            return
        self.width = width
        self.height = height
        self.reset()

    # --- commands from the player ----------------------------------------

    def start(self) -> None:
        if self.status in (GameStatus.READY, GameStatus.PAUSED):
            self.status = GameStatus.RUNNING

    def pause(self) -> None:
        if self.status is GameStatus.RUNNING:
            self.status = GameStatus.PAUSED

    def toggle_pause(self) -> None:
        """Space is a single key, so pausing and resuming are the same command."""
        if self.status is GameStatus.PAUSED:
            self.status = GameStatus.RUNNING
        else:
            self.pause()

    def turn(self, direction: Direction) -> None:
        """A turn also starts the game, so the first arrow key just works."""
        if self.status is GameStatus.READY:
            self.status = GameStatus.RUNNING
        if self.status is GameStatus.RUNNING:
            self.snake.turn(direction)

    # --- the loop ---------------------------------------------------------

    def tick(self) -> None:
        if self.status is not GameStatus.RUNNING:
            return

        target = self.snake.next_head()
        if is_fatal(target, self.snake.cells, self.width, self.height):
            self.status = GameStatus.GAME_OVER
            return

        # Resolve food *before* moving, so growth and score land on the same
        # tick. Growing first also keeps the tail in place, which is what makes
        # the new segment appear immediately rather than one tick late.
        eating = self.food is not None and self.food == target
        if eating:
            self.snake.grow()

        self.snake.move()
        self.ticks += 1

        if eating:
            self.score += 1
            # Every apple flips the whole world to the next pair, in order.
            self.palette = (self.palette + 1) % PALETTE_COUNT
            if not self._respawn_food():
                self.status = GameStatus.GAME_OVER  # board full: the snake won

    # --- what the browser sees --------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        """The wire format. Keep in sync with web/types/game.ts."""
        return {
            "type": "state",
            "width": self.width,
            "height": self.height,
            "status": self.status.value,
            "score": self.score,
            "ticks": self.ticks,
            "direction": self.snake.direction.name,
            "snake": [[x, y] for x, y in self.snake.cells],
            "food": list(self.food) if self.food is not None else None,
            "palette": self.palette,
        }

    # --- internals ---------------------------------------------------------

    def _respawn_food(self) -> bool:
        """Place food on a free cell. False when the board is full.

        Samples from the *free* cells rather than retrying random guesses, so a
        nearly-full board still terminates.
        """
        free = [
            (x, y)
            for y in range(self.height)
            for x in range(self.width)
            if not self.snake.occupies((x, y))
        ]
        if not free:
            self.food = None
            return False
        self.food = self._rng.choice(free)
        return True

File: game-server/src/Game.cpp
#include "Game.hpp"

#include <algorithm>
#include <random>
#include <string>
#include <vector>

#include <nlohmann/json.hpp>

#include "Collision.hpp"

namespace snake {

std::string_view nameOf(GameStatus status) {
    switch (status) {
        case GameStatus::Ready:    return "ready";
        case GameStatus::Running:  return "running";
        case GameStatus::Paused:   return "paused";
        case GameStatus::GameOver: return "game_over";
    }
    return "ready";
}

Game::Game(int width, int height, double tickSeconds, std::optional<std::uint32_t> seed)
    : width_(width),
      height_(height),
      tickSeconds_(tickSeconds),
      rng_(seed ? *seed : std::random_device{}()),
      snake_(Cell{width / 2, height / 2}) {
    reset();
}

void Game::reset() {
    snake_ = Snake(Cell{width_ / 2, height_ / 2}, Direction::Right);
    food_.reset();
    respawnFood();
    score_ = 0;
    ticks_ = 0;
    status_ = GameStatus::Ready;
}

void Game::start() {
    if (status_ == GameStatus::Ready || status_ == GameStatus::Paused) {
        status_ = GameStatus::Running;
    }
}

void Game::pause() {
    if (status_ == GameStatus::Running) status_ = GameStatus::Paused;
}

void Game::togglePause() {
    if (status_ == GameStatus::Paused) {
        status_ = GameStatus::Running;
    } else {
        pause();
    }
}

void Game::turn(Direction direction) {
    if (status_ == GameStatus::Ready) status_ = GameStatus::Running;
    if (status_ == GameStatus::Running) snake_.turn(direction);
}

void Game::tick() {
    if (status_ != GameStatus::Running) return;

    const Cell target = snake_.nextHead();
    if (isFatal(target, snake_.cells(), width_, height_)) {
        status_ = GameStatus::GameOver;
        return;
    }

    // Resolve food *before* moving, so growth and score land on the same tick.
    // Growing first also keeps the tail in place, which is what makes the new
    // segment appear immediately rather than one tick late.
    const bool eating = food_.has_value() && *food_ == target;
    if (eating) snake_.grow();

    snake_.move();
    ++ticks_;

    if (eating) {
        ++score_;
        if (!respawnFood()) {
            status_ = GameStatus::GameOver;  // board full: the snake won
        }
    }
}

bool Game::respawnFood() {
    const std::deque<Cell>& body = snake_.cells();

    std::vector<Cell> free;
    free.reserve(static_cast<std::size_t>(width_) * height_);
    for (int y = 0; y < height_; ++y) {
        for (int x = 0; x < width_; ++x) {
            const Cell cell{x, y};
            if (std::find(body.begin(), body.end(), cell) == body.end()) {
                free.push_back(cell);
            }
        }
    }

    if (free.empty()) {
        food_.reset();
        return false;
    }

    std::uniform_int_distribution<std::size_t> pick(0, free.size() - 1);
    food_ = free[pick(rng_)];
    return true;
}

std::string Game::toJson() const {
    nlohmann::json out{
        {"type", "state"},
        {"width", width_},
        {"height", height_},
        {"status", nameOf(status_)},
        {"score", score_},
        {"ticks", ticks_},
        {"direction", nameOf(snake_.direction())},
    };

    nlohmann::json body = nlohmann::json::array();
    for (const Cell& cell : snake_.cells()) {
        body.push_back({cell.x, cell.y});
    }
    out["snake"] = std::move(body);

    if (food_) {
        out["food"] = {food_->x, food_->y};
    } else {
        out["food"] = nullptr;
    }

    return out.dump();
}

}  // namespace snake
